### PR TITLE
Past bans

### PR DESCRIPTION
**Pull request by stonedDiscord**
this keeps past bans in the db so you can look up for what someone got banned in the past

~~it also shows displays the remaining ban time to the user in a more human readable format
in 20 minutes instead of todays date~~ that was my other PR

the WIP stuff is the database, i need to remove some UNIQUE constraints

### CODE REVIEW

**Comment by oldmud0** [comment]
Will merge after 2.8 compatibility is finished and stable.

**Comment by stonedDiscord** [comment]
bump

File: migrations/v4.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	ban_id INTEGER NOT NULL,
+	FOREIGN KEY (ban_id) REFERENCES bans(ban_id)
+		ON DELETE CASCADE
+);
+INSERT INTO ip_bans_new SELECT * FROM ip_bans;
+DROP TABLE ip_bans;
+ALTER TABLE ip_bans_new RENAME TO ip_bans;
+
+-- Add unbanned column
+ALTER TABLE bans ADD unbanned INTEGER DEFAULT 0;
```
Comment (oldmud0) [inline migrations/v4.sql:15]: Hmm. Wondering if it makes more sense to invert this column, e.g. `active`, or to use the unban date directly to figure out if the ban is still in effect.

**Review by oldmud0 — COMMENTED** [review]
Some minor issues:
- `/baninfo` won't retrieve a full ban history. Might need to add an index to `ooc_events` to speed up OOC name lookup.
- Attaching an extra IPID/HDID to an existing ban using the `/ban <ipid> <ban_id>` syntax will not warn if that ban is no longer in force.

**Comment by stonedDiscord** [comment]
i specifically don't want to use the unban date to invalidate a ban so you can keep track of who was banned for how long, even after they get unbanned by the moderators

i will work on the minor issues you listed soon

File: server/commands/admin.py [hunk 2]
```diff
@@ -396,21 +396,24 @@ def ooc_cmd_baninfo(client, arg):
...
-            msg += f'Unban date: {unban_date.format()} ({unban_date.humanize()})'
-        else:
-            msg += 'Unban date: N/A'
+        msg = f'Bans for {args[0]}'
+        for ban in bans:
+            msg += f'\nBan ID: {ban.ban_id}\n'
+            msg += 'Affected IPIDs: ' + ', '.join([str(ipid) for ipid in ban.ipids]) + '\n'
+            msg += 'Affected HDIDs: ' + ', '.join(ban.hdids) + '\n'
+            msg += f'Reason: "{ban.reason}"\n'
+            msg += 'Unbanned: {}\n'.format(bool(ban.unbanned))
```
Comment (oldmud0) [inline server/commands/admin.py]:
```suggestion
            msg += f'Unbanned: {bool(ban.unbanned)}\n'
```

File: server/database.py [hunk 3]
```diff
@@ -277,6 +277,26 @@ def find_ban(self, ipid=None, hdid=None, ban_id=None):
...
+            bans = conn.execute(dedent('''
+                SELECT *
+                FROM (
+                    SELECT ban_id FROM ip_bans WHERE ipid = ?
+                    UNION SELECT ban_id FROM hdid_bans WHERE hdid = ?
+                    UNION SELECT ban_id FROM bans WHERE ban_id = ?
+                )
+                JOIN bans USING (ban_id)
+                '''), (ipid, hdid, ban_id)).fetchall()
+            if bans is not None:
```
Comment (oldmud0) [inline server/database.py]: `bans` will be `[]` not `None` when empty, so this if-statement is useless

File: migrations/v4.sql (hunk 1 above)
Comment (stonedDiscord) [reply migrations/v4.sql:15]: https://github.com/AttorneyOnline/tsuserver3/pull/98#issuecomment-745641340

**Comment by stonedDiscord** [comment]
since #161 also increments the DB version we should combine the migration scripts

**Comment by oldmud0** [comment]
I'm fine with this as long as it was well tested.

**Pull request merged**